### PR TITLE
Hacky work around for bug apparent in Samsung Android 14 with SurfaceView

### PR DESCRIPTION
**Pull request by dnfield**
This does not have a test because the only real way to reproduce the issue it's addressing is currently flaky and requirest a device type we do not have in CI.

See comments for repro steps.

I have tried a number of times locally to reproduce, and with this work around have not been able to do so.

For future me: This reproduces less often in unopt builds.

https://github.com/flutter/flutter/issues/139630

### CODE REVIEW

**Review by johnmccutchan — APPROVED** [review]

File: shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java [hunk 1]
```diff
@@ -56,6 +58,35 @@ public void surfaceCreated(@NonNull SurfaceHolder holder) {
 
           if (shouldNotify()) {
             connectSurfaceToRenderer();
+
+            // Workaround for https://github.com/flutter/flutter/issues/139630
+            // We should not need this if Samsung fixes b/316626640
```
Comment (Hixie) [inline shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java:63]: "b/316626640" is a Google-internal-ism.

File: shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java [hunk 2]
```diff
@@ -56,6 +58,35 @@ public void surfaceCreated(@NonNull SurfaceHolder holder) {
...
+            // This requires painful manually testing to reproduce.
+            // Run the default counter application on a Samsung phone running
+            // on Anroid 14. Background the application. Wait a few minutes.
```
Comment (Hixie) [inline shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java:66]: typo: Android

File: shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java [hunk 3]
```diff
@@ -56,6 +58,35 @@ public void surfaceCreated(@NonNull SurfaceHolder holder) {
...
+
+                        @Override
+                        public void doFrame(long frameTimeNanos) {
+                          requestLayout();
+                          count--;
+                          if (count > 0) {
+                            Choreographer.getInstance().postFrameCallback(this);
+                          }
+                        }
+                      });
```
Comment (Hixie) [inline shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java:88]: Are we sure we want to do this? This is basically saying that Samsung devices get handled differently than other devices. This is going to lead to situations where things can reproduce in one device but not another.

Can we do something like detect the failure mode and do the hack when we detect the failure mode, instead?

**Comment by Hixie** [comment]
Is there an issue for this problem?

**Comment by johnmccutchan** [comment]
> Is there an issue for this problem?

https://github.com/flutter/flutter/issues/139630

File: shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java (hunk 3 above)
Comment (dnfield) [reply shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java:88]: We're currently in a situation where things reproduce in one device but not another.

We cannot detect the failure mode here - the places where the failure is visible are all internal to SurfaceFlinger and not accessible from Java or C++ based app code on Android.

File: shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java (hunk 2 above)
Comment (dnfield) [reply shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java:66]: Done

File: shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java (hunk 1 above)
Comment (dnfield) [reply shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java:63]: Yes - this bug is only visible to Googlers and (some?) Samsung employees unfortunately. I've updated the text around it, but I'd like to keep it here for people who have access.

**Comment by chinmaygarde** [comment]
We are going to hold on landing this for now. There are no updates on the internal issue.

File: shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java (hunk 1 above)
Comment (Yokoi-K) [reply shell/platform/android/io/flutter/embedding/android/FlutterSurfaceView.java]: @Hixie @dnfield 

I would like to know what the current progress/status is regarding this Google internal issue.
"b/316626640"

Can you please let me know?

**Comment by dnfield** [comment]
Someone from Samsung has been able to reproduce this bug. I'd like to wait a few more days to see if they have any fixes or better ideas for fixing this on their end.

**Comment by chinmaygarde** [comment]
Samsung has not responded with anything meaningful on this issue via internal channels. We are considering landing this to prevent the pain.

**Comment by Hixie** [comment]
I strongly recommend not landing this. Every experience i've had with this kind of hack is telling me that if we start down this path we will regret it deeply.

(It would be less problematic if we were to always apply the hack, rather than only on some devices.)

**Comment by dnfield** [comment]
Samsung is telling us they've fixed this bug and it should show up for users soon. I'm going to close this.

@Hixie  - I agree with you, but FWIW we definitely have other vendor specific hacks in code we own and in code we consume (Skia has a lot of this).

**Comment by Hixie** [comment]
> @Hixie - I agree with you, but FWIW we definitely have other vendor specific hacks in code we own and in code we consume (Skia has a lot of this).

I would encourage us to remove such hacks, or at a very minimum come up with a coherent strategy (e.g. a centralized database that the code consulted to enable/disable hacks rather than hard-coding checks, with a switch one could flip that turns off all the hacks, or reports which hacks are being used, etc... but really not having any vendor-specific hacks is much preferred).

There's a variety of problems with vendor-specific hacks:
 - they prevent the vendors from fixing their problems (because now they have to be compatible with our hacks)
 - they balloon our testing matrix (to the point where we definitely aren't testing every combination)
 - they make it harder to develop (because different devices do different things)
 - they make it harder to document (because we have to explain all the variations)
 - they introduce new bugs (because these are hacks against essentially undocumented behavior and so we're unlikely to get precisely the right logic to counter that behavior)
 - etc

**Pull request closed without merging**